zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Learner.java

```diff
@@ -756,13 +760,21 @@ protected void syncWithLeader(long newLeaderZxid) throws Exception {
                     zk.startupWithoutServing();
                     if (zk instanceof FollowerZooKeeperServer) {
                         FollowerZooKeeperServer fzk = (FollowerZooKeeperServer) zk;
-                        for (PacketInFlight p : packetsNotCommitted) {
+                        fzk.syncProcessor.setDelayForwarding(true);
+                        for (PacketInFlight p : packetsNotLogged) {
                             fzk.logRequest(p.hdr, p.rec, p.digest);
                         }
-                        packetsNotCommitted.clear();
+                        packetsNotLogged.clear();
```

Comment:
is this one of the key bugs? did we need to keep packetsNotCommitted around?

Reply:
This was the bug that would cause the learner to crash during sync, because it "forgot" a previous PROPOSAL on the NEWLEADER, and would then fail to match up that PROPOSAL with a later COMMIT, if one was sent during the sync. 
In turn, this caused the learner to have to re-sync, which could trigger the same crash again if there was heavy concurrent write traffic, and it would also give duplicate series of transactions in the transaction logs, with resulting transaction digest mismatch on that server (but otherwise consistent data view). 

So we need a separation of what's not yet written to the log, and what's not yet matched with a COMMIT, which is what these two queues are about.